### ISSUE TITLE
Next-word-predictions development discussion thread

### ISSUE THREAD

**Issue by patrickgold** [florisboard/florisboard#325]
Hi all!

The time has finally come, I am working on dictionary suggestions! As I want the development of this feature to be as transparent and open as possible, I will share the roadmap for the next 2-3 weeks with you. This issue thread serves as a room for general discussion about the implementation and feedback given in this issue will directly influence the early development of suggestions. The development of word suggestions is rather big, so I split it up in four main phases. After finishing each phase, the already implemented part of the suggestions will be released, so all users on F-Droid and Google Play can try it out. Only if you want to, you can also install the debug artifact builds which get created for each commit in the PR, so you can try out new stuff even faster.

During development, all preferences regarding this topic will be marked as [EXPERIMENTAL] and all string resources regarding this feature will be non-translatable. I am doing this to prevent circumstances where I change a string like 5 times and so the translators get notified 5 times to translate the same string over and over again, even though often only a small thing has changed. Nearing the end of the last phase, I will unlock these strings for translation.

For each phase beginning from phase 1, a separate PR will be filed and linked here.

NOTE: below phases are currently crossed out because the time plan got completely overthrown by the OOM issues as well as the Flictionary implementation not working out as it should. I don't have a concrete plan for suggestions, but I think I will implement spellchecking (and suggesting spell-checked words) first, as this part is a bit easier to add than next-word predictions. The old Flictionary implementation will stay in for now until I've found a replacement.

### Phase 0
~This phase has technically already finished. It includes collecting ideas for which sources can be used (a big thanks to everyone who commented on #34!!), providing a way to have standardized dictionary formats in the APK (see https://github.com/florisboard/dictionary-tools) as well as reading and understanding how word suggestions should work. As this phase didn't change a thing in the application itself, no release for this phase is being made.~

### Phase 1 (v0.3.8) (#329)
~This is the "real" first phase, in which the input logic code gets an overhaul. This especially changes the `EditorInstance` logic to be more performant while providing better context to the prediction logic. Also, simple tests are conducted if the created binary dictionaries from the dictionary-tools are actually readable on Android and are built correctly. Then, the logic for providing words for the current input is created. Note, that the predictions in this phase are quite "dumb", as the next-word algorithm is missing.~

~Supported languages in this phase:~
- ~English~

### Phase 2 (v0.3.11)
~In this phase, experiments regarding the next-word algorithms are conducted and the decision is made, whether to use a traditional "Algorithm" or a local AI module. Additionally, more languages (currently though general only, no specific region dialects etc.) are supported:~
- ~English~
- ~German~
- ~Spanish~
- ~Portuguese~
- ~Italian~
- ~French~

### Phase 3 (v0.3.12)
~- Further improvement of the prediction and next-word algorithm.~
- ~All languages from https://github.com/LuminosoInsight/wordfreq are supported, still though no region/dialect specific sources.~

### Phase 4 (v0.3.13)
~- Final bug-fixing~
- ~Translations get unlocked on Crowdin for the suggestion-related UI strings.~
- ~If requested, I can look into adding languages from other sources rather than LuminosoInsight as well in this phase to support region/dialect specific word predictions.~

~After Phase 4, Word suggestions for the "main" languages should be up and running. This does of course doesn't mean that word suggestions are finished. I will then continuously add new languages, add support for dialects of languages, improve the prediction algorithm, fix bugs, etc.~

~I hope that this roadmap gives you a clear insight on what I'll be working on in the next few weeks. Also note that especially in the first week I will not work on other open bug reports/feature requests (except when a critical security bug arises which has to be fixed asap), so I can fully concentrate on suggestions.~

If you have any thoughts about this or want to give feedback, feel free to comment below!

**Comment by Kreuger** [comment]
You are the man!

**Comment by kj7rrv** [comment]
Would it be possible to support unsupported languages by learning as the
user types?

--... ...-- / -.. . / -.- .--- --... .-. .-. ...-

**Comment by patrickgold** [comment]
@kj7rrv That's pretty difficult, as I would start out with nothing and it would take a long time to train an AI model to provide useful suggestions for that unsupported language. It is always good to have a decent wordlist as a base to start out.

**Comment by Glitchy-Tozier** [comment]
> currently though general only, no specific region accents etc.

What does that mean?
Will users only be able to insert "Tur" instead of "Tür"?

**Comment by patrickgold** [comment]
> What does that mean?

I just noticed that I meant dialects/region specific language and not accents 🤦 Will correct this in my comment. So for instance if you are a user speaking `de_AT`, you'd still only get suggestions for the "main" `de` language pack, no specialized suggestions. And "Tür" certainly is a normal German word which is included in the main `de` language pack :)

**Comment by Glitchy-Tozier** [comment]
> I just noticed that I meant dialects/region specific language and not accents 🤦 Will correct this in my comment. So for instance if you are a user speaking `de_AT`, you'd still only get suggestions for the "main" `de` language pack, no specialized suggestions. And "Tür" certainly is a normal German word which is included in the main `de` language pack :)

Okay, that's better! I was scared for a moment you'd not allow any diacritics 😅

**Comment by sabzo** [comment]
@patrickgold for the dictionary-tools are you using [this project](https://github.com/remi0s/aosp-dictionary-tools) or its logic?

This is really neat, and just some thoughts...
### For Learning on the Fly
regarding @kj7rrv's question, learning without an existing model or wordlist has been done by several projects 
which implement _User History_. AnySoftKeyboard, OpenBoard, [Indic Keyboard](https://github.com/smc/Indic-Keyboard/blob/89ba9f517ece85b5b3d367042e5ed6b9ba4a0667/tests/src/com/android/inputmethod/latin/personalization/UserHistoryDictionaryTests.java). 

* The indic keyboard is to unit tests that use the User History feature.

The potential challenge is that all the implementations are Java and rely on C++ functionality using JNI interface -- based on the original Android Open Source keyboard code. 

### Prediction Algorithms

1. **Word Prediction** N-grams was suggested, which is good. Indic/OpenBoard use bi-grams, Swiftkey a few years ago used by tri-grams. There's a helpful article on [N-grams](https://towardsdatascience.com/nlp-with-kotlin-ec84c068650) in Kotlin along with the [source code](https://github.com/mccorby/MachineLearning/tree/master/src/main/kotlin/com/mccorby/machinelearning/nlp).
2. **Neural Nets** It would be useful to leave the api open-ended so that developers may replace the underlying algorithm, for example being able to swap out a bi-gram with a quad-gram model, or maybe a bi-gram with a Neural Network model.

### Spell Check algorithm
For Spellcheck/Autocorrect There's a [very good article](https://towardsdatascience.com/symspellcompound-10ec8f467c9b) and [implementation ](https://github.com/wolfgarbe/symspell) and it seems to be the real deal. There isn't a Kotlin version nevertheless. 

Again, neat stuff happening here!

**Comment by patrickgold** [comment]
> for the dictionary-tools are you using this project or its logic?

@sabzo I wanted to use this when I planned to use the AOSP dict format, but now I am using the Flictionary format which requires its own logic (built by me in Python).

> regarding @kj7rrv's question, learning without an existing model or wordlist has been done by several projects
which implement User History. AnySoftKeyboard, OpenBoard, Indic Keyboard.

Interesting to see, I will look into it (at least the concept, because the Java / C++ / JNI mix of code is a bit confusing to read)

> Word Prediction N-grams was suggested, which is good. Indic/OpenBoard use bi-grams, Swiftkey a few years ago used by tri-grams. There's a helpful article on N-grams in Kotlin along with the source code.

That's currently what I am experimenting and trying to understand how it works. What exactly I am gonna be using is currently undecided.

> Neural Nets It would be useful to leave the api open-ended so that developers may replace the underlying algorithm, for example being able to swap out a bi-gram with a quad-gram model, or maybe a bi-gram with a Neural Network model.

I want to use such a model for next-word-prediction which gets trained by the user and after some time can reliably predict the next-word.

> For Spellcheck/Autocorrect There's a very good article and implementation and it seems to be the real deal. There isn't a Kotlin version nevertheless.

Thanks for linking, will look into it.

**Comment by sabzo** [comment]
> That's currently what I am experimenting and trying to understand how it works. What exactly I am gonna be using is currently undecided.

I stumbled upon the project recently, and am trying to get a good grasp of it, but I could proof the n-gram prediction code, or contribute somehow, I've done a good amount of work in NLP from n-grams to RNNs.

> I want to use such a model for next-word-prediction which gets trained by the user and after some time can reliably predict the next-word. 

Alright it seems you're focusing on the [User Dictionary](https://developer.android.com/reference/kotlin/android/provider/UserDictionary.Words) the one trained by a user over time. To clarify, would developers pass in their own dictionaries (flictionaries) to give users some AI functionality from the start?

**Comment by patrickgold** [comment]
> I stumbled upon the project recently, and am trying to get a good grasp of it, but I could proof the n-gram prediction code, or contribute somehow, I've done a good amount of work in NLP from n-grams to RNNs.

Reviewing the code of n-gram would definitely be cool, because especially at the beginning I don't really know where to start and the result may look questionable from an expert's eye, so any feedback is appreciated :)

> Alright it seems you're focusing on the User Dictionary the one trained by a user over time. To clarify, would developers pass in their own dictionaries (flictionaries) to give users some AI functionality from the start?

Well I am currently making a deep-dive into the fields of NLP and n-gram and I am currently experimenting with a bi-gram runtime implementation for next-words. The Flictionaries which are included are there to assist the _current word_ prediction based on the already inputted chars (and maybe the spellcheck too but I'm not sure), but do not assist in the _next word_ prediction, as a unigram list is kinda useless for this I've discovered. For next word, I will use bi/tri gram models which will get trained by the user over time while writing, maybe I will pre-train it with some open corpus, so it has a headstart when first used.

**Comment by sabzo** [comment]
> Well I am currently making a deep-dive into the fields of NLP and n-gram and I am currently experimenting with a bi-gram runtime implementation for next-words. The Flictionaries which are included are there to assist the _current word_ prediction based on the already inputted chars (and maybe the spellcheck too but I'm not sure), but do not assist in the _next word_ prediction, as a unigram list is kinda useless for this I've discovered. For next word, I will use bi/tri gram models which will get trained by the user over time while writing, maybe I will pre-train it with some open corpus, so it has a headstart when first used.

The latest commit doesn't have the _current_ word prediction yet, but I'd be curious to see what algorithm you used afterall.

> Well I am currently making a deep-dive into the fields of NLP and n-gram and I am currently experimenting with a bi-gram runtime implementation for next-words. The Flictionaries which are included are there to assist the _current word_ prediction based on the already inputted chars (and maybe the spellcheck too but I'm not sure), but do not assist in the _next word_ prediction, as a unigram list is kinda useless for this I've discovered. For next word, I will use bi/tri gram models which will get trained by the user over time while writing, maybe I will pre-train it with some open corpus, so it has a headstart when first used.

Exactly, a unigram list by definition isn't next word prediction (as it's looking at only one word at a time, where as n-gram you're looking at the previous n-1 words), and your custom UserDictionary will have to build its own bigramg/trigram list. Default AOSP supports a bigram list.  You may as well do trigrams, which will be an improvement over the AOSP, with the knowledge trigrams will require more space. 

But please kindly make the _current word prediction_ and _next word prediction_ all abstractable so that others can add/substitute their own nlp implementations. The difficulty with the current open source Android keyboard is that the code isn't modular -- which requires too much investigation and trial and error just to safely replace some functionality.

**Comment by patrickgold** [comment]
> The latest commit doesn't have the current word prediction yet, but I'd be curious to see what algorithm you used afterall.

Yeah i know because I am currently going around in circles and don't really produce something commit-worthy. Also i am just reading and experimenting with the concepts, I will though hopefully tomorrow be able to push a basic framework for this.

> Exactly, a unigram list by definition isn't next word prediction (as it's looking at only one word at a time, where as n-gram you're looking at the previous n-1 words), and your custom UserDictionary will have to build its own bigramg/trigram list. Default AOSP supports a bigram list. You may as well do trigrams, which will be an improvement over the AOSP, with the knowledge trigrams will require more space.

My idea for next-word is to use a mix of bi and tri gram, whereas trigrams are the preferred and bigrams are fallback/supportive additional.

> But please kindly make the current word prediction and next word prediction all abstractable so that others can add/substitute their own nlp implementations. The difficulty with the current open source Android keyboard is that the code isn't modular -- which requires too much investigation and trial and error just to safely replace some functionality.

If you look at the way I began to implement the Dictionary, you can see that it is a standardized interface. All methods in the code base will only work with the interface. The actual implementation of the dictionary is the Flictionary. This would theoretically allow to add other types of Binary dictionaries, as long as it implements the Dictionary interface. The same approach do I want to take in the prediction related stuff. + I don't mix Java / C++ / JNI because that makes everything worse in terms of readability, everything is 100% Kotlin.

**Comment by sabzo** [comment]
> If you look at the way I began to implement the Dictionary, you can see that it is a standardized interface. All methods in the code base will only work with the interface. The actual implementation of the dictionary is the Flictionary. This would theoretically allow to add other types of Binary dictionaries, as long as it implements the Dictionary interface. The same approach do I want to take in the prediction related stuff. + I don't mix Java / C++ / JNI because that makes everything worse in terms of readability, everything is 100% Kotlin.

Hmmm, I imagine for others to move from the other AOSP keyboards they would have to create an adapter from aosp dict to flictionary dict, and the adapter to implement Flictionary's Dictionary interface. Inevitably the adapter would probably be a re-write or port of some java/jni/c++ -- unless if there was a way where someone can supply the [AOSP wordlist](https://android.googlesource.com/platform/packages/inputmethods/LatinIME/+/android-sdk-4.4.2_r1/dictionaries/sample.combined) which will then convert to a Flictionary binary.

In other words the intervention is probably best at the wordlist level than at the binary dictionary level?





> My idea for next-word is to use a mix of bi and tri gram, whereas trigrams are the preferred and bigrams are fallback/supportive additional.

Sounds good.

> Yeah i know because I am currently going around in circles and don't really produce something commit-worthy. Also i am just reading and experimenting with the concepts, I will though hopefully tomorrow be able to push a basic framework for this.

Looking forward to helping where possible

**Comment by patrickgold** [comment]
> unless if there was a way where someone can supply the AOSP wordlist which will then convert to a Flictionary binary.

Flictionary binaries are built from combined lists, see this method:
https://github.com/florisboard/dictionary-tools/blob/main/flict.py#L96-L126

`src_path` is the combined list, `dst_path` the path of the Flictionary binary. The only thing is, that it currently does not support the `bigram` field or `shortcut`, only the generic `word` and `f`.

As I've seen that the AOSP dictionary supports `bigram`, does that mean that an AOSP dictionary binary can theoretically house a bigram model? If so, it means that I'd have to expand the Flictionary format spec or even rethink it.

**Comment by sabzo** [comment]
> > unless if there was a way where someone can supply the AOSP wordlist which will then convert to a Flictionary binary.
> 
> Flictionary binaries are built from combined lists, see this method:
> https://github.com/florisboard/dictionary-tools/blob/main/flict.py#L96-L126
> 
> `src_path` is the combined list, `dst_path` the path of the Flictionary binary. The only thing is, that it currently does not support the `bigram` field or `shortcut`, only the generic `word` and `f`.
> 
> As I've seen that the AOSP dictionary supports `bigram`, does that mean that an AOSP dictionary binary can theoretically house a bigram model? If so, it means that I'd have to expand the Flictionary format spec or even rethink it.

You're exactly right, an AOSP dictionary does support bigram and in fact does house bigram models. I've built several bigram binary dictionary files that do just that, simply by adding the _bigram_ keyword and an associated frequency. Also the AOSP format can support words that shouldn't be recommended, although the spelling is correct (maybe there may be offensive or crude) -- it's an interesting format because so much can be specified. The only limitation being that it doesn't support an arbitrary "n"-gram, so tri-gram would be an improvement, or the ability to add any N-gram. 

> Flictionary binaries are built from combined lists, see this method:
> https://github.com/florisboard/dictionary-tools/blob/main/flict.py#L96-L126

Awesome, thanks for sharing that.

**Comment by patrickgold** [comment]
I have an idea how to rework the Flictionary format spec so it can house any n-gram space efficiently.

What do you think of this reworked spec for the Flictionary: https://github.com/florisboard/dictionary-tools/blob/flict-spec-rework/flictionary.md
It allows to save a unigram word list like before as a ptree and optionally any n-grams as a ptree. The N is not bound, so 2, 3, 4, ...- grams can be saved all in one while saving storage space, as long as the word has been indexed in the unigram list before.

**Comment by sabzo** [comment]
> What do you think of this reworked spec for the Flictionary: https://github.com/florisboard/dictionary-tools/blob/flict-spec-rework/flictionary.md
> It allows to save a unigram word list like before as a ptree and optionally any n-grams as a ptree. The N is not bound, so 2, 3, 4, ...- grams can be saved all in one while saving storage space, as long as the word has been indexed in the unigram list before.

Great, the N not being bound is an **immediate** improvement on other AOSP keyboards! For a N-gram word to already exist in the unigram list is fine and would cover a majority of the use cases (recommending the nth word if it is not in a unigram).

I can see how the compression you used 24 bits versus 8 bit * sizeof(word) would make sense for small file size -- which means you're probably thinking of storing User History Dictionary as a flictionary file? Which would minimized storage for user history dict.

Noting that on some AOSP keyboard the User History Dictionary is built as a binary I _believe_ each time the Keyboard View is open -- curious if the flicitonary binaries also take advantage of some compression. The dict binaries end up being the ones taking up most of the space.

**Comment by patrickgold** [comment]
Thanks for your detailed feedback, I really appreicate it! :)

> an unlikely case maybe recommending the nth word if it is not in a unigram 

This is an edge case, but still an issue in the format. One solution could be to say: freq=0 -> potentially offensive word (like AOSP implemented this), freq=1 -> not a word but required for n-grams, freq = 2..255 -> normal valid word.

> I can see how the compression you used 24 bits versus 8 bit * sizeof(word) would make sense for small file size -- 

It gets even worse for non-Latin script languages, where a character may be 3 or 4 bytes in length. This is why I chose to index the words, so I have them exactly once in my file to save storage space.

> which means you're probably thinking of storing User History Dictionary as a flictionary file? Which would minimized storage for user history dict.

Yep, I plan to have at minimum 3 runtime binaries in the cache:

Prebuilt EN | UserHistory EN | Personal
:-----------:|:----------------:|:--------:
read-only | read/write | read/write
This dictionary is the one which gets built with the combined-list and which also causes the discussion on how to deliver it to the end user (#339). One dictionary for each language is built here and then distributed. | This dictionary is the local user dictionary, one for each subtype used by the user. Gradually while typing, the unigram as well as the n-gram model grows and refines itself. This dictionary never leaves the User's device (except it is manually exported, then it is up to the user what happens further). | This dictionary provides names, email addresses, etc. and is the same for all languages. Also never leaves the user's device except when exported.

The lookup priority for the suggestion algorithm is right-to-left. Each Dictionary can have a `LanguageModel` attached to it, but this is not required. Next-word suggestions work only with LanguageModels existing though, as the algorithm needs bi/trigrams to have a base for prediction. Probably only the UserHistory dictionary will have such model attached, as it makes most sense to save the learned data there.

> I'm also wondering if you're taking into account the ability to create multi-language autocomplete/prediction? There are cases where users may want to have auto-complete or word-prediction in 2 or 3 languages without having to switch the subtype, some people may write in multiple languages during the same conversation, or same sentence.

I am not 100% sure about this one, but I think loading multiple prebuilt and user history dictionaries (one for each subtype) should be a base for multi-lingual support.

This afternoon I also worked on the abstract implementation of the Dictionaries and the NLP skeleton code and have just uploaded the abstract implementation to the suggestions branch. If possible, could you have a look at it and give me feedback if it somewhat makes sense and that I didn't mess up the naming?

https://github.com/florisboard/florisboard/tree/2b001d9eb8521a93a30941a0b3a6419df5594a0b/app/src/main/java/dev/patrickgold/florisboard/ime/nlp (all files except `TextProcessor`)
and
https://github.com/florisboard/florisboard/blob/2b001d9eb8521a93a30941a0b3a6419df5594a0b/app/src/main/java/dev/patrickgold/florisboard/ime/dictionary/Dictionary.kt

**Comment by sabzo** [comment]
> This is an edge case, but still an issue in the format. One solution could be to say: freq=0 -> potentially offensive word (like AOSP implemented this), freq=1 -> not a word but required for n-grams, freq = 2..255 -> normal valid word.

Ahhh, creative use of freq=1.  As long as there would be a separate keyword, kinda like how shortcut is used as a keyword for 0 - 14 in the AOSP wordlists. 

> It gets even worse for non-Latin script languages, where a character may be 3 or 4 bytes in length. This is why I chose to index the words, so I have them exactly once in my file to save storage space.
Got it.

> Probably only the UserHistory dictionary will have such model attached, as it makes most sense to save the learned data there.

Yes, I think that's how it is elsewhere. Awesome use as well to abstract LanguageModel!

> I am not 100% sure about this one, but I think loading multiple prebuilt and user history dictionaries (one for each subtype) should be a base for multi-lingual support.

Exactly, being able to load N dictionaries (but truthfully the quality or utility begins to degrade after 3 or 4) -- diminishing returns; Gboard currently allows up to 3 language, some keyboards just 2.



> https://github.com/florisboard/florisboard/tree/2b001d9eb8521a93a30941a0b3a6419df5594a0b/app/src/main/java/dev/patrickgold/florisboard/ime/nlp (all files except TextProcessor)
and
https://github.com/florisboard/florisboard/blob/2b001d9eb8521a93a30941a0b3a6419df5594a0b/app/src/main/java/dev/patrickgold/florisboard/ime/dictionary/Dictionary.kt

Now looking at the code, I'll respond a bit later. Although you haven't done the implementation yet, a quick heads up on frequency. In some cases the bigram's frequency may be lower than it's unigram score -- in that case the bigram is disregarded when building the dictionary -- the AOSP jar tool does this functionality. The benefit of a bigram should be that the bigram gets a higher f score as it is more likely than its original unigram position.

Also some stop words may end up having high f scores (and, the, but, of ) etc., and it can get annoying in the unigram lists, and these words may also show up highly ranked in the n-gram list --but that's something NLP practitioners can replace, which is why the LanguageModel abstraction is just great. 

... gonna look at code and respond

**Comment by patrickgold** [comment]
Thanks for your feedback!

So I've rethought the Flictionary spec once again, this is the (hopefully final) structure I propose:
![image](https://user-images.githubusercontent.com/19412843/108073935-e96bd500-7068-11eb-8043-bbb84e64b8b7.png)

Each Flictionary consists of 3 areas. The header hasn't changed and still contains its meta data as before. Then there's the address tree. Its purpose is to index all words which occur in this Flictionary formatted as a ptree and thus save them space-efficiently. The difference to the previous approach is, that the address tree now does not provide any NLP related properties anymore, so any words can be indexed here.

The third and most important section is the n-gram tree or the language model. In it, all n-grams (unigram, bigram etc.) are stored. The most important part here is that each node only knows the token address which it describes, not the token itself while packed in the binary file, which allows me to save huge amounts of storage space. Each n-gram tree node can be of either types:
- **word**: This n-gram node marks the end of a valid n-gram and attaches a frequency to it. A freq value of 0 indicates possibly offensive (only the last node though), every other value can be freely used (1-255).
- **shortcut**: May only be used for unigram nodes and marks that this word itself is not a valid word, through it can point to another word. Useful for things such as comand -> command etc.
- **filler**: If this node does not mark the end of a n-gram, but is required in a higher order n-gram.

Now the question is are there any other types the AOSP dictionary can store which I am not aware of? As I am still in the early phase of the Flictionaries, it is easy now to add such things than later.

My proposed change change would not require that much rewrite in the abstract implementation (except for Token). When unpacking the model into runtime, the addresses will get replaced by the actual words. When packing again, the words have to be re-indexed.

> Now looking at the code, I'll respond a bit later. Although you haven't done the implementation yet, a quick heads up on frequency. In some cases the bigram's frequency may be lower than it's unigram score -- in that case the bigram is disregarded when building the dictionary -- the AOSP jar tool does this functionality. The benefit of a bigram should be that the bigram gets a higher f score as it is more likely than its original unigram position.

That is something I don't understand. So lets say we have this:

```
 word=sample,f=200
  bigram=wordlist,f=170
 word=wordlist,f=180
```

which evaluates to this:

```
sample 200
sample wordlist 170
wordlist 180
```

Why should the bigram be discarded? It means that "wordlist" itself occurs with a frequency of 180 but a bit less when being preceded by "sample". I see no reason here to disregard it because it is still useful information for the next-word algorithm.

> Also some stop words may end up having high f scores (and, the, but, of ) etc., and it can get annoying in the unigram lists, and these words may also show up highly ranked in the n-gram list --but that's something NLP practitioners can replace, which is why the LanguageModel abstraction is just great.

This is true. When looking at the first 20 entries of the English wordlist, almost exclusively stopwords occur. I can't take them out though because if I use the same unigram list for spellchecking, the spellchecker would mark these as invalid (as there's no entry for them).

**Comment by sabzo** [comment]
> Now the question is are there any other types the AOSP dictionary can store which I am not aware of? As I am still in the early phase of the Flictionaries, it is easy now to add such things than later.

you've covered every condition specified by [AOSP](https://android.googlesource.com/platform/packages/inputmethods/LatinIME/+/android-sdk-4.4.2_r1/dictionaries/sample.combined) 

> My proposed change change would not require that much rewrite in the abstract implementation (except for Token). When unpacking the model into runtime, the addresses will get replaced by the actual words. When packing again, the words have to be re-indexed.

Fair.

> Why should the bigram be discarded? It means that "wordlist" itself occurs with a frequency of 180 but a bit less when being preceded by "sample". I see no reason here to disregard it because it is still useful information for the next-word algorithm.

I didn't mean to be confusing, in your example,  _sample wordlist 170_ will never get recommended because _wordlist 180_ has a higher precedence, even after the unigram _sample_ has been written. In the binary creation process by the aosp-tool jar it "throws away" that line entirely, I assume to save space. Keeping the bigram _sample wordlist_ is redundant as it will never be actually called at all. In this case the language model cares about predicting the sequence with the highest probabilities. 

Stated differently:

Let's keep the bigram _sample wordlist 170_. If there's another bigram _sample wordset 165_ neither of those bigrams will get recommended even if the first part of the bigram exist all because _wordlist 180_ is greater than all the potential _bigrams_. The utility/usefulness of the bigram or a language model is that it bumps the end word (using NLP terminology a language model takes n-1 words to predict the nth word or nth token to be more precise). Using this definition, if having _sample_ doesn't boost the likelihood of _wordlist), or _wordset_ then there's no need storing the bigrams _sample wordlist_ and _sample wordset_. Instead wordlist/wordset are better of as unigrams since they're more likely to occur in that context.

For ex: in some of my wordlists, the most frequent words are probably around 180, then the bigrams f score is > 200 and the score becomes higher as there are less choices of the bigram to choose from.

otherwise the format looks good and the diagram looks good and I'll think will be clear to others who have use AOSP combined list format! Awesome.

**Comment by sabzo** [comment]
@patrickgold on the picture you posted is there an implied hierarchy represented by tabs/spacing (it looks like you tabbed some boxes)? The hierarchy that seems to fit is with _filler_ and _word_ combinations.

**Comment by patrickgold** [comment]
> you've covered every condition specified by AOSP

Perfect :)

> I didn't mean to be confusing, in your example,  _sample wordlist 170_ will never get recommended because _wordlist 180_ has a higher precedence, even after the unigram _sample_ has been written.

Ok I think I am missing some crucial part in understanding how to use n-gram models for text prediction. Sorry if I am not the brightest in this topic but my brain just refuses to accept this, even after reading a few documents and articles about NLP n-gram language models. I understand the methods to analyze a text with n-grams, but for predicting text I just don't get it :(

So what I thought how the language model was used is this (not saying this way is good or correct, just what I thought is good):

```
[Language model attached to UserDictionary]
the 220
the sample 180
car 150
the car 160
the model 145
the sun 70
sample 200
sample wordlist 170
sample wordset 165
sequence 10
wordlist 180
wordlist is 138
wordset 130
```

Now the user has inputted the following thus far (`|` is the input cursor): `the |`

The current token `w` is empty, the history token list `h` is `["the"]`. This means I don't know anything about the next word, except the history. So I am searching the language model for the following bigrams: `("the", *)`, where `*` is a wildcard for any token. The returned list would be the following:

```
the sample 180
the car 160
the model 145
the sun 70
```

Here you can see, that the list returned ("the" "car" 160), even though "car" alone is 150. Why? Because the wildcard search required to have a "the" before any word, and the unigram didn't have that. Based on this, I can take the first three entries ("sample", "car" and "model") because they have the highest frequency and show these in the suggestion bar. Now the user presses a character and the input looks like this: `the s|`

Now I can be assured that the next word begins with a `s`. First, I am searching my already fetched list if a token begins with a `s`. In our case, only "sample" for now. Then, I am searching all bigrams for the following: `("the", "s"*)` and take all entries (except the already fetched "sample"). Now I have only 2 out of 3 needed suggestions, so I will look in the unigram list if it can help me out here, with the highest freq counter being "sequence". My three suggestions for this round are now ("sample", "sun" and "sequence"). And this process gets repeated for every word.

In this way of creating predictions, it makes no sense for me to leave a lower frequency bigram out, because the lookup process' priority is always higher to lower order. Now the problem is that this approach is probably very inefficient or does not provide that good results what I thought and there are way better ways to predict words, where unigrams are involved and thus the lower freq bigram makes no sense. The hard part for me is to understand exactly this and how the unigram for the next word is prioritized even though the user hasn't even typed the word yet.

> For ex: in some of my wordlists, the most frequent words are probably around 180, then the bigrams f score is > 200 and the score becomes higher as there are less choices of the bigram to choose from.

When following the logic that higher-order n-grams must have higher frequencies, this makes perfect sense to me, as well as the fact that the AOSP jar leaves this bigrams out to save storage space and because it is redundant in that case. The hard part for me is to understand why the unigram is prioritized over the bigram and not the other way round in next word predictions.

Another question I have now is if the n-grams should also include `<start_of_sentence>` and `<end_of_sentence>` tokens? I have seen this tokens used quite a lot in n-gram NLP articles and documents and it seems that they are often included, but not always.

Thanks for your patience till now and for your efforts in discussing this with me, I really appreciate it :)

**Comment by patrickgold** [comment]
> on the picture you posted is there an implied hierarchy represented by tabs/spacing (it looks like you tabbed some boxes)? The hierarchy that seems to fit is with filler and word combinations.

Yes the indentation by spacing should represent the level in the n-gram tree we are in. If we have this language model:

```
the 220
the sample 180
the moon 96
the sun 128
car model 150
car model color 145
```

this gets saved to the Flictionary in this way to save storage space:

```
the f=220
    sample f=180
    moon f=96
    sun f=128
car [filler, no f attached]
    model f=150
        color f=145
```

No indentation means unigram, 1 indentation means bigram, 2 trigram etc.

**Comment by sabzo** [comment]
No worries at all @patrickgold.
> The hard part for me is to understand why the unigram is prioritized over the bigram and not the other way round in next word predictions.

This is a mistake on my part! -- I've also reran Open Board's AOSP wordlist and tested the prediction algorithm in order give this info: If there are at least 2 words: bigrams predictions take precedent if they are there **even** If there are higher frequency unigrams. It was difficult to understand because it was incorrect, you had the right intuition. 

> Now I have only 2 out of 3 needed suggestions, so I will look in the unigram list if it can help me out here, with the highest freq counter being "sequence". My three suggestions for this round are now ("sample", "sun" and "sequence"). And this process gets repeated for every word.

This may not always be true -- just because you began by typing `s` doesn't mean the autocomplete will display words that begin with s. For example in my keyboard if I type `s` I as `As` and `Is` as likely predictions. Suppose s is uncommon in the subtype you're in, maybe it may show words that begin with say `z` or something. Sometimes the word auto-completed maybe based from the "edit distance". In your example you're no longer doing "next word prediction" per-se, you're performing _auto-complete_.

> Another question I have now is if the n-grams should also include <start_of_sentence> and <end_of_sentence> tokens? I have seen this tokens used quite a lot in n-gram NLP articles and documents and it seems that they are often included, but not always.

tl;dr I favor adding the "padding" tokens as necessary in order to meet the "N" quorum. 

I'd recommend [This 10 minute video from Dan Jurafasky at Stanford](https://www.youtube.com/watch?v=KO5JdmPjtRo&list=PLLssT5z_DsK8HbD2sPcUIDfQ7zmBarMYv&index=15). 
Yes, \<bos>/<eos` is one implementation and it works. The easiest option is to treat everything like a higher dimension N-Gram, say tri-gram. But there could be a case where you have more unigrams, and bigrams, than trigrams. The trigrams maybe sparse. Suppose you take the vocabulary of all english words and try to find sequences of 3 words. There will most likely be sequences far fewer sequences of 3 words than say two words. If unigrams, bigrams, and trigrams are all represented the same way `LanguageModel3DArray[first][second][third] = some probablity` there maybe a a lot of zero probabilities (just like video states). In that case one can use "stupid backoff" a form of "smoothing" so as to apply trigrams where they best apply, otherwise use the lower dimension "grams".

If you want to treat all grams as a set N-gram, then you have to put `<start-token>` etc., If you want trigrams you have to put this twice `<start-token><start-token><actual-word>`. The goal is to have three tokens in order to satisfy the sequence of n words rule. The first two start words are there to help satisfy have a sequence of 3 words. But some of these things you would worry more about them if you were building "machine translation" or "spell checkers". Next word prediction is an easier case, because you're predicting words you already know. Whereas spelling you may find a N-gram sequence you have never seen and you have to validate if it's correct or not, and perhaps same with MT. 

All to say I don't see too much harm in using a set "N" grams for all words, whether a word may be a unigram, part of a bigram, trigram etc., if it's to predict the next word. For example converting each word into a trigram sequence. For other "downstream" nlp tasks more care may be needed in dealing with "smoothing" (as video suggests). At the very least it would be good enough to keep going, and if there are other more nuanced approaches, others could add them since the api would be modular. 

If quality is an issue, bear in mind that in many cases using Neural Networks would probably generate significantly better word predictions -- the rabbit hole doesn't stop! N-grams are a strong start, especially being able to deal with higher order n-grams. I don't think you'll have to worry about _smoothing_ or switching between say trigrams, bigrams etc., I think that process should be abstracted.

**Comment by Glitchy-Tozier** [comment]
I don't understand most of the things discussed here but this
> First, I am searching my already fetched list if a token begins with a s

seems a little redundant if you are going to do this
> Then, I am searching all bigrams for the following: ("the", "s"*) and take all entries (except the already fetched "sample").

anyways.

___

> Another question I have now is if the n-grams should also include <start_of_sentence> and <end_of_sentence> tokens?

Is this a question whether you should respect the start of sentences at all or shether this is the correct implementation?
If it's the first, i feel like the answer is yes. If it's the second, no idea.

____

> If quality is an issue, bear in mind that in many cases using Neural Networks would probably generate significantly better word predictions

Wasn't this what you intended to do @patrickgold ?

**Comment by patrickgold** [comment]
> This is a mistake on my part! -- I've also reran Open Board's AOSP wordlist and tested the prediction algorithm in order give this info: If there are at least 2 words: bigrams predictions take precedent if they are there even If there are higher frequency unigrams. It was difficult to understand because it was incorrect, you had the right intuition.

Ah ok good to know, thanks for testing it out :) Although I am still far away from fully understanding the topic, I now at least understand it good enough to be able to try writing an implementation for this and do learn-by-doing  for  the rest.

> This may not always be true -- just because you began by typing s doesn't mean the autocomplete will display words that begin with s. For example in my keyboard if I type s then As and Is are shown as likely predictions. Suppose s is uncommon in the subtype you're in, maybe the edit distance or some measure of word similarity, may show words that begin with say z or something. Sometimes the word auto-completed maybe based from the "edit distance". In your example you're no longer doing "next word prediction" per-se, you're performing auto-complete.

That's true, I absolutely didn't thought of that. I think working with the edit distance here should give me good results for the next word. I will start with the Kotlin implementation tomorrow (technical college kept me a bit too busy the last few days to work on FlorisBoard) and also rewrite the Flictionary encoder/decoder to work with the reworked file structure.

> tl;dr I favor adding the "padding" tokens as necessary in order to meet the "N" quorum.

K, that's what I thought too. Thanks also for the detailed explanation for these special tokens, I now have a good picture of the usefulness and why they should be included.

> First, I am searching my already fetched list if a token begins with a s
seems a little redundant if you are going to do this
Then, I am searching all bigrams for the following: ("the", "s"*) and take all entries (except the already fetched "sample").
anyways.

Yup, thinking about it now this step does not make sense at all. But like I said I will have to work with an edit distance measure to provide more accurate results, so this approach as I described it won't be included in the final result.

> Is this a question whether you should respect the start of sentences at all or shether this is the correct implementation?
If it's the first, i feel like the answer is yes. If it's the second, no idea.

Yes this was a question whether I should include these special tokens at all, and I think all three of us agree that these are a nice-to-have and don't cause any harm to the usefulness of the language model.

> Wasn't this what you intended to do @patrickgold ?

@Glitchy-Tozier Yes, I planned (and still think of) using neural networks for this task. For now, I will see where the n-gram model takes me, and as @ sabzo said - the rabbit hole never ends in this topic, so I can definitely think of using a combi of both or only neural networks in the future. For now, it is important for me to fully understand how NLP works without neural networks, so I have a good base for everything further.

**Comment by sabzo** [comment]
@patrickgold No worries! I think you have all the pieces to get moving, I'm excited for this feature!

**Comment by Glitchy-Tozier** [comment]
> For now, it is important for me to fully understand how NLP works without neural networks, so I have a good base for everything further.

Makes sense!

**Comment by MisterY** [comment]
This raises very interesting questions:

> This may not always be true -- just because you began by typing `s` doesn't mean the autocomplete will display words that begin with s. For example in my keyboard if I type `s` then `As` and `Is` are shown as likely predictions. Suppose `s` is uncommon in the subtype you're in, maybe the edit distance or some measure of word similarity, may show words that begin with say `z` or something. Sometimes the word auto-completed maybe based from the "edit distance". In your example you're no longer doing "next word prediction" per-se, you're performing _auto-complete_.

I.e. should the keyboard try to compensate for the "fat-fingers", where the small phone keyboards make it easy to type the nearby character. And which characters are nearby depends on the selected layout.

**Comment by sabzo** [comment]
Earlier you gave this "language model" -- 

> the 220
> the sample 180
> the moon 96
> the sun 128
> car model 150
> car model color 145

Is this the type of file that's passed to the python3 `dict-tool`?

And From the previous example, how does one get to this -- is this the input to `dict-tool`?

> the f=220
>     sample f=180
>     moon f=96
>     sun f=128
> car [filler, no f attached]
>     model f=150
>         color f=145

Luminoso](https://github.com/LuminosoInsight/wordfreq) doesn't have actual examples of the generated files -- only python commands to run -- even then, it doesn't account for n-grams > 1. 

Can you show a simple example, maybe a wordlist of 10 lines showing unigram/bigram that will act as input to the `dict-tool` -- it maybe helpful for others to see the pipeline of generating "flicts". 


Thanks!

**Comment by patrickgold** [comment]
> Earlier you gave this "language model" -- Is this the type of file that's passed to the python3 dict-tool?

No, that was just an example to showcase a language model for explaining here on GitHub. The actual input format of `flict.py` is a AOSP-style combined list.

> And From the previous example, how does one get to this -- is this the input to dict-tool?

Nope, also this was just for visualization.

> Luminoso](https://github.com/LuminosoInsight/wordfreq) doesn't have actual examples of the generated files -- only python commands to run -- even then, it doesn't account for n-grams > 1.

It defines word lists in the `cBpack` file format (see [this link](https://github.com/LuminosoInsight/wordfreq/blob/4c0b29f460d32e81582c0b201270b56a0cf9cf6a/wordfreq/__init__.py#L37-L76) which is the best description I've found in the source code). This format is then converted to an AOSP style combined list in `clb.py`.

In general, this is how the dicttool is designed to work:
- Step 1: Convert the given input format to a standardized AOSP combined list in `clb.py`. For every new data source we support, we have to add a new converter method.
- Step 2: In `flict.py`, we take the previously generated AOSP combined wordlist and make a Flictionary binary file out of this. If your input is already in the AOSP style format, you can skip step 1.

![image](https://user-images.githubusercontent.com/19412843/109270260-101ad000-780e-11eb-9ff9-6479d3079675.png)

> Can you show a simple example, maybe a wordlist of 10 lines showing unigram/bigram that will act as input to the dict-tool -- it maybe helpful for others to see the pipeline of generating "flicts".

Here's an example of the input combined list for `flict.py`:

```
dictionary=main:en,locale=en,description=Auto-generated dictionary for en,date=1614300627744,version=1
 word=the,f=255
  bigram=word,f=230
   trigram=list,f=91
  bigram=sun,f=200
  bigram=car,f=190
   trigram=wash,f=80
 word=of,f=254
 word=to,f=254
 word=and,f=254
 word=a,f=254
  bigram=word,f=200
 word=in,f=254
 word=i,f=254
 word=is,f=254
 offensiveword,f=0
  ...
```

Important to note is that `shortcut` and `is_not_a_word` are currently not recognized by the converter, also anything further beyond trigrams. These will be added in a later stage.

**Comment by Glitchy-Tozier** [comment]
Btw, should there be the same suggestion three times?

(The thing I typed was "ommmmmmmmmmmm")

![Screenshot_20210303-112730_Phone.jpg](https://user-images.githubusercontent.com/59611881/109792801-2eb90680-7c14-11eb-8a8d-b83795e8d91d.jpg)

**Comment by patrickgold** [comment]
@Glitchy-Tozier No, there shouldn't. I am aware that this can happen, I have to prevent the suggestion list from accepting already existing words.

**Comment by ftyers** [comment]
Have you considered using a character-based RNN language model to make suggestions. I think Android has TensorflowLite support built in. There is some code that can be probably reused here: https://github.com/apertium/apertium-keyboard. I would be interested in working on it.

**Comment by patrickgold** [comment]
@ftyers In a way or another I already use a character based model currently, see https://github.com/florisboard/florisboard/blob/master/app/src/main/java/dev/patrickgold/florisboard/ime/nlp/FlorisLanguageModel.kt

The code of the keyboard you linked looks indeed very usable, as it uses TensorFlow lite to do the heavy memory&computation stuff, which currently are one of the main concerns.

> I would be interested in working on it.

If you want to experiment with adding some of the code to FlorisBoard, you can!

I have a few questions to TensorFlow though:
- Are the tflite models for word suggestions pre-generated for each supported language and then learn while the user types or does the model start with zero for each user?
- TensorFlow [lite] is open-source and has an Apache 2.0 license, though it is still a Google product, thus I am always a bit cautious at first when it comes to Google's privacy practices. As far as I understood TensorFlow lite runs 100% locally and requires no Internet permission, so all data learned from typing stays within the app, right? It is important that this is ensured because the whole purpose of FlorisBoard is to put user's privacy at first.
- What size does the TensorFlow lite library has when added to the APK? For many FlorisBoard users it is important that the APK size stays relatively small, thus I am asking beforehand.

**Comment by sabzo** [comment]
@ftyers I tried to run the code, it seems it's stuck on Tensorflow 1, whereas the latest version of TF is version 2. 

@patrickgold FL lite is the gold standard for mobile app models or browser based. Since it's open source I would put more trust to it. 

I went on a goose chase with https://github.com/attractivechaos/kann a small c lightweight neural network library. I got it to run on FlorisBoard using the NDK however getting predictions for a word took about 400-500 ms. If one needed 3 word choices it's more than a second -- so I abandoned it for now. To try to get more optimization it would need refactoring, however the code is poorly documented.

There's also Pytorch as well for Android but it's hectic to say the least.

**Comment by ftyers** [comment]
> @ftyers In a way or another I already use a character based model currently, see https://github.com/florisboard/florisboard/blob/master/app/src/main/java/dev/patrickgold/florisboard/ime/nlp/FlorisLanguageModel.kt
> 
> The code of the keyboard you linked looks indeed very usable, as it uses TensorFlow lite to do the heavy memory&computation stuff, which currently are one of the main concerns.
> 
> > I would be interested in working on it.
> 
> If you want to experiment with adding some of the code to FlorisBoard, you can!
> 
> I have a few questions to TensorFlow though:
> 
>     * Are the tflite models for word suggestions pre-generated for each supported language and then learn while the user types or does the model start with zero for each user?

You can generate your own models per language and then it can learn from the user. I would probably just generate some model with OpenSubtitles or something like that to start with. Model training is fairly straightforward.

>     * TensorFlow [lite] is open-source and has an Apache 2.0 license, though it is still a Google product, thus I am always a bit cautious at first when it comes to Google's privacy practices. As far as I understood TensorFlow lite runs 100% locally and requires no Internet permission, so all data learned from typing stays within the app, right? 

That is correct. I also dislike Google, but unfortunately to my knowledge there are few available on-device libraries for neural networks. You could look into Pytorch, but then that's FaceBook. Or you could write your own, Tensorflow is a lot of overhead, but you'd need to look into how all the GPU integration stuff is done, running stuff on CPU is likely to be fairly slow.

>It is important that this is ensured because the whole purpose of FlorisBoard is to put user's privacy at first.

That's why I'm here.

>     * What size does the TensorFlow lite library has when added to the APK? For many FlorisBoard users it is important that the APK size stays relatively small, thus I am asking beforehand.

No idea, but it's probably shipped with quite a few apps.

**Comment by ftyers** [comment]
> @ftyers I tried to run the code, it seems it's stuck on Tensorflow 1, whereas the latest version of TF is version 2.

That is correct, I have not had time to port it to Tensorflow 2. I actually really loathe working with Tensorflow, the tooling is awful, bazel is a pain, etc. 

> @patrickgold FL lite is the gold standard for mobile app models or browser based. Since it's open source I would put more trust to it.
> 
> I went on a goose chase with https://github.com/attractivechaos/kann a small c lightweight neural network library. I got it to run on FlorisBoard using the NDK however getting predictions for a word took about 400-500 ms. If one needed 3 word choices it's more than a second -- so I abandoned it for now. To try to get more optimization it would need refactoring, however the code is poorly documented.

That's some interesting code, I've often thought about doing something similar. 

> There's also Pytorch as well for Android but it's hectic to say the least.

Also run by Facebook, so has the same issues as Google regarding privacy.

**Comment by sabzo** [comment]
@ftyers 

> You can generate your own models per language and then it can learn from the user. I would probably just generate some model with OpenSubtitles or something like that to start with. Model training is fairly straightforward.

How do you do the on device learning with tf-lite? I thought once it's on device only inference is happening rather than learning. If learning can indeed happen it would be useful for user experience!

**Comment by mgrove36** [comment]
Hi all, I've just discovered this project and love it! Some great work here. Having read through the relevant issues I can't seem to find explicit mentions of autocorrection (replacing the typed word with the first suggestion when the spacebar is pressed) instead of just providing suggestions. This may well just be me missing something, but could you provide some clarification please?
I'm loving this project already; well done everyone!

**Comment by sabzo** [comment]
@mgrove36 The issue with that is that pressing a spacebar isn't necessary an indicator that the user wanted the first word suggested. This could be a new word not in the dictionary. Perhaps in a future version if the user retypes the word that was autocorrected due to spacebar press the word would be added to the user history. FlorisBoard currently doesn't support/have user history.

**Comment by mgrove36** [comment]
Ok, thanks for the response! Loving what's happening here :)

**Comment by Glitchy-Tozier** [comment]
@patrickgold Maybe we should remove/change the version-numbers next to **Phase X**. They are a little outdated.

**Comment by papjul** [comment]
Hi, I tried Florisboard, found out there was no French dictionary and saw this issue.

Having participated in an update of the French dictionary in ASK, I thought I could help here, especially since the implementation in Florisboard looks very similar.

In French, the most popular and maintained option for dictionaries is Grammalecte-dic (previously known as Dicollecte), which is used (among others) in Hunspell.

You can download the "Lexique" file here: https://grammalecte.net/download.php?prj=fr
It contains a .csv file (inside a .txt, don't ask me why), with the column "Indice de fréquence" (frequency index) being a number from 0 to 9.

Since it's too limited for our use case (numbers are between 0 and 255), we may use "Fréquence" instead, which is the actual frequency of the word (in Wikipedia pages I think?).
However, if we do it like this, we will have the top ten words being between let's say 10 and 255, and the others words sharing the last bottom numbers.
So we need some kind of exponential function here to smooth the curve of frequency, or alternatively put aside the top ten words and set them up as 255 and use a linear function for the other words.

I had done a Python script to handle this, but lost it unfortunately :(
However I'm willing to go back to it if it can help having a great FOSS alternative to glide keyboards x)
Let me know if you need more information or any help implementing it, as I would be very happy to see it happening soon ;)

**Comment by linusrg1** [comment]
Unless i jist missed it, could you possibly add an option to make your keyboard one solid colour throughout the os. As well as that punless I missed this in the settings too, could you add an option to remove the big "English" text from the spacebar?

**Comment by Glitchy-Tozier** [comment]
@papjul Currently focus is on removing the constant crashes. Once that's done (in 1-2 weeks), devellopment should shift back to other things such as suggestions.

@linusrg1 Please don't post feature-requests in random  discussions. Instead, open a new proposal-issue and describe your thoughts there.
(One issue per proposal)

**Comment by papjul** [comment]
Thank you for your reply ;) No worries, take your time fixing the nasty bugs!

I think that "suggestions thread" title is a bit misleading, it can be understood as "discuss suggestions to improve the app here".
Maybe it can be renamed "New feature: Word suggestions" or something like this ;)

**Comment by Glitchy-Tozier** [comment]
> I think that "suggestions thread" title is a bit misleading, it can be understood as "discuss suggestions to improve the app here".

Ooooooh I never thought of that issue!  😅

@patrickgold , opinion?

**Comment by patrickgold** [comment]
> Ooooooh I never thought of that issue! sweat_smile

Me neither tbh. But yeah, for someone not familiar with the current issues and topics this can indeed sound a bit misleading. Will rename this thread then.

**Comment by sabzo** [comment]
> @papjul Currently focus is on removing the constant crashes. Once that's done (in 1-2 weeks), devellopment should shift back to other things such as suggestions.

What are the inclinations so far? Will this be Rust, C/C++, or a small neural model (pytorch/tensorflow)?

**Comment by patrickgold** [comment]
> What are the inclinations so far? Will this be Rust, C/C++, or a small neural model (pytorch/tensorflow)?

@sabzo In the last few weeks I thought about this and it will most likely be going towards C++ mixed with the usage of a neural network. Reasons for this are that for spellchecking I want to utilize Hunspell, which is already in C++. Also Kotlin + C++ with JNI interfaces is somewhat documented, Kotlin + Rust not so much. The next word predictions will use a mix of some pre-compiled dictionaries and a dynamical locally trained network to predict words (with TensorFlow lite most probably). Pytorch is pretty much crossed out, because of performance on mobile devices.

**Comment by KaKi87** [comment]
Hello, everything is crossed out yet I have no non-english support on FlorisBoard v0.3.12

**Comment by Glitchy-Tozier** [comment]
@patrickgold Maybe this issue should be unpinned/closed/rewritten, as some things have changed.

**Comment by patrickgold** [comment]
> Hello, everything is crossed out yet I have no non-english support on FlorisBoard v0.3.12

Everything is crossed out because word suggestions (the first implementation) was a bit of chaos and caused a lot of crashes. Since then the crashes have been resolved, though development on this particular feature has grinded to a halt. Currently thus only English (US) is available. In the feature roadmap you can see that I plan to re-do word suggestions in 0.3.15 (currently I am prepping 0.3.13), so this is the version I am aiming to have suggestions for all languages, similar to the fully implemented FlorisBoard system spell checker service.

> Maybe this issue should be unpinned/closed/rewritten, as some things have changed.

I agree, because this issue is currently not relevant. I will probably also open a separate new discussion issue for the 0.3.15 suggestions in the future, so I will additionally close this discussion thread.

**Comment by KaKi87** [comment]
So, where is this tracked now, and when can I hope for my language to be supported ?

**Comment by patrickgold** [comment]
The whole suggestions thing is not really tracked rn, the single issues/bug reports and feature requests for word suggestions are all tracked though under the `area: word-prediction` label and in the feature roadmap.

> when can I hope for my language to be supported ?

In 0.3.15. No ETA for this though because I don't know how long the 0.3.14 preference re-write (currently beginning to work on this one) and extension base support will take.
